velox/functions/prestosql/aggregates/ReservoirSampleAggregate.cpp

```diff
@@ -0,0 +1,663 @@
+        nullptr,
+        numGroups,
+        initialSampleOffsets,
+        initialSampleSizes,
+        initialSampleElementsVector);
+
+    *result = std::make_shared<RowVector>(
+        pool,
+        rowType,
+        nullptr,
+        numGroups,
+        std::vector<VectorPtr>{
+            sampleVector,
+            initialSampleVector,
+            initialSeenCountVector,
+            seenCountVector,
+            maxSampleSizeVector});
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    VELOX_CHECK_EQ(args.size(), 4, "reservoir_sample requires 4 arguments");
+
+    DecodedVector initialSampleDecoded(*args[0], rows);
+    DecodedVector initialProcessedCountDecoded(*args[1], rows);
+    DecodedVector valuesToSampleDecoded(*args[2], rows);
+    DecodedVector desiredSampleSizeDecoded(*args[3], rows);
+
+    auto elementType = args[2]->type();
+
+    rows.applyToSelected([&](vector_size_t row_idx) {
+      auto* accumulator = value<ReservoirSampleAccumulator>(groups[row_idx]);
+
+      if (!accumulator->isInitialized()) {
+        initializeAccumulator(
+            accumulator,
+            initialSampleDecoded,
+            initialProcessedCountDecoded,
+            desiredSampleSizeDecoded,
+            row_idx,
+            elementType);
+      }
+
+      // Add new value using reservoir sampling
+      accumulator->processedCount++;
+      accumulator->addValueToReservoir(
+          valuesToSampleDecoded, row_idx, elementType);
+    });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    VELOX_CHECK_EQ(args.size(), 1, "intermediate results requires 1 argument");
+
+    DecodedVector decodedIntermediate(*args[0], rows);
+    auto* rowVector = decodedIntermediate.base()->as<RowVector>();
+    auto* sampleArrayVector = rowVector->childAt(0)->as<ArrayVector>();
+    auto* initialSampleArrayVector = rowVector->childAt(1)->as<ArrayVector>();
+    auto* initialSeenCountVector =
+        rowVector->childAt(2)->asFlatVector<int64_t>();
+    auto* seenCountVector = rowVector->childAt(3)->asFlatVector<int64_t>();
+    auto* maxSampleSizeVector = rowVector->childAt(4)->asFlatVector<int32_t>();
+
+    auto elementType = sampleArrayVector->elements()->type();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (decodedIntermediate.isNullAt(row)) {
+        return;
+      }
+
+      auto* accumulator = value<ReservoirSampleAccumulator>(groups[row]);
+      auto decodedRow = decodedIntermediate.index(row);
+
+      int32_t maxSampleSize = maxSampleSizeVector->valueAt(decodedRow);
+      if (maxSampleSize < 0) {
+        return;
+      }
+
+      if (!accumulator->isInitialized()) {
+        new (accumulator) ReservoirSampleAccumulator(allocator_->pool());
```

Comment:
accumulator is still a valid object right, so why recreate a placement new on previously created accumulator ? Might this not be dangerous ?